### PR TITLE
Type cache in program for bn type conversion

### PR DESCRIPTION
**Pull request by kumarak**
Add `TypeCache` to maintain the program level cache for binja to anvill type conversion instead of using the local cache for each type.

### CODE REVIEW

File: python/anvill/binja.py [hunk 1]
```diff
@@ -173,199 +173,260 @@ def _convert_bn_llil_type(
...
-        ret = TypedefType()
-        cache[_cache_key(ref_type)] = ret
-        ret.set_underlying_type(_convert_bn_type(bv, ref_type, cache))
         return ret
 
-    elif (named_tinfo.type_class
-        == bn.NamedTypeReferenceClass.EnumNamedTypeClass):
-        # Set the underlying type int of size width
-        ref_type = bv.get_type_by_name(named_tinfo.name);
+    def __convert_enum(self, tinfo: bn.types.Type) -> Type:
```
Comment (surovic) [inline python/anvill/binja.py:243]: I think we should decide if we want to use type hints in the code or not. I have a slight preference for having them, but I don't really mind either way. What's important to me is consistency.

**Review by surovic — CHANGES_REQUESTED** [review]
Just a heads up, remember to run a formatter on the code. I use `black` in default setting, which I think is PEP8?

**Comment by kumarak** [comment]
> Just a heads up, remember to run a formatter on the code. I use `black` in default setting, which I think is PEP8?

Thanks for the heads up. I was not sure of the code formatting setup. I will run black with the default setting.

File: python/anvill/binja.py [hunk 2]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
         return ret
 
-    elif (named_tinfo.type_class
-        == bn.NamedTypeReferenceClass.EnumNamedTypeClass):
-        # Set the underlying type int of size width
-        ref_type = bv.get_type_by_name(named_tinfo.name);
+    def __convert_enum(self, tinfo: bn.types.Type):
+        """Convert bn enum type into a `Type` instance"""
+
+        if tinfo.type_class != bn.TypeClass.EnumerationTypeClass:
```
Comment (pgoodman) [inline python/anvill/binja.py:246]: `assert`

File: python/anvill/binja.py [hunk 3]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
+        return MacOS()
+    elif "windows" in platform:
+        return WindowsOS()
+    else:
+        raise UnhandledOSException(
+            "Missing operating system object type for OS '{}'".format(platform)
+        )
+
+
 def _cache_key(tinfo: bn.types.Type) -> str:
```
Comment (pgoodman) [inline python/anvill/binja.py:207]: Move into `TypeCache`.

File: python/anvill/binja.py [hunk 4]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
-        struct_type =  ref_type.structure
+class TypeCache(object):
+    def __init__(self, bv):
+        self._bv = bv
+        self._cache = dict()
+
+    def __convert_struct(self, tinfo: bn.types.Type):
+        """Convert bn struct type into a `Type` instance"""
+
+        if tinfo.structure.type != bn.StructureType.StructStructureType:
```
Comment (pgoodman) [inline python/anvill/binja.py:220]: `assert`

File: python/anvill/binja.py [hunk 5]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
-        == bn.NamedTypeReferenceClass.StructNamedTypeClass):
-        # Get the bn struct type and recursively recover the elements
-        ref_type = bv.get_type_by_name(named_tinfo.name);
-        struct_type =  ref_type.structure
+class TypeCache(object):
+    def __init__(self, bv):
+        self._bv = bv
+        self._cache = dict()
+
+    def __convert_struct(self, tinfo: bn.types.Type):
```
Comment (pgoodman) [inline python/anvill/binja.py:217]: Single underscore prefix on these private methods.

File: python/anvill/binja.py [hunk 6]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
 
-    elif (named_tinfo.type_class
-        == bn.NamedTypeReferenceClass.UnionNamedTypeClass):
-        # Get the union type and recover the member elements
-        ref_type = bv.get_type_by_name(named_tinfo.name);
-        struct_type =  ref_type.structure
+    def __convert_union(self, tinfo: bn.types.Type):
+        """Convert bn union type into a `Type` instance"""
+
+        if tinfo.structure.type != bn.StructureType.UnionStructureType:
```
Comment (pgoodman) [inline python/anvill/binja.py:233]: assert

File: python/anvill/binja.py [hunk 7]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
+            return
+
         ret = UnionType()
-        cache[_cache_key(struct_type)] = ret
-        for elem in struct_type.union.members:
+        cache[_cache_key(tinfo)] = ret
+        for elem in tinfo.structure.union.members:
             ret.add_element_type(_convert_bn_type(bv, elem.type, cache))
```
Comment (pgoodman) [inline python/anvill/binja.py:239]: `self. _convert_bn_type(elem.type)`

File: python/anvill/binja.py [hunk 8]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
+        # tinfo.width
+        ret.set_underlying_type(IntegerType(tinfo.width, False))
         return ret
 
-    else:
-        DEBUG("WARNING: Unknown named type {} not handled".format(named_tinfo))
+    def __convert_typedef(self, tinfo: bn.types.Type):
+        """ Convert bn typedef into a `Type` instance"""
 
+        if tinfo.type_class != bn.NamedTypeReferenceClass.TypedefNamedTypeClass:
```
Comment (pgoodman) [inline python/anvill/binja.py:259]: assert

File: python/anvill/binja.py [hunk 9]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
+        cache[_cache_key(tinfo)] = ret
+        ret.set_underlying_type(
+            self._convert_bn_type(self._bv.get_type_by_name(tinfo.name))
+        )
+        return ret
+
+    def __convert_array(self, tinfo: bn.types.Type):
+        """ Convert bn pointer type into a `Type` instance"""
+
+        if tinfo.type_class != bn.TypeClass.ArrayTypeClass:
```
Comment (pgoodman) [inline python/anvill/binja.py:272]: assert

File: python/anvill/binja.py [hunk 10]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
-        # The underlying type of enum will be an Interger of size
-        # tinfo.width
-        ret = EnumType()
-        cache[cache_key] = ret
-        ret.set_underlying_type(IntegerType(tinfo.width, False))
-        return ret
-
-    elif tinfo.type_class == bn.TypeClass.BoolTypeClass:
-        return BoolType()
+        if tinfo.type_class != bn.TypeClass.IntegerTypeClass:
```
Comment (pgoodman) [inline python/anvill/binja.py:313]: assert

File: python/anvill/binja.py [hunk 11]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
-        ret.set_element_type(_convert_bn_type(bv, tinfo.element_type, cache))
+        self._cache[_cache_key(tinfo)] = ret
+        ret.set_element_type(self._convert_bn_type(tinfo.element_type))
         return ret
 
-    elif tinfo.type_class == bn.TypeClass.FunctionTypeClass:
+    def __convert_function(self, tinfo):
+        """ Convert bn function type into a `Type` instance"""
+
+        if tinfo.type_class != bn.TypeClass.FunctionTypeClass:
```
Comment (pgoodman) [inline python/anvill/binja.py:295]: assert

File: python/anvill/binja.py [hunk 12]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
+        elif tinfo.type_class == bn.TypeClass.IntegerTypeClass:
+            return self.__convert_integer(tinfo)
+
+        elif tinfo.type_class == bn.TypeClass.FloatTypeClass:
+            return FloatingPointType(tinfo.width)
+
+        elif tinfo.type_class == bn.TypeClass.NamedTypeReferenceClass:
+            return self.__convert_named_reference(tinfo)
+
+        elif tinfo.type_class in [
```
Comment (pgoodman) [inline python/anvill/binja.py:386]: Make the `err_type_class` dictionary a global var, then check if `tinfo.type_class` is in the dict, i.e. is a key in the dict.

File: python/anvill/binja.py [hunk 13]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
+        if isinstance(ty, Type):
+            return ty
+
+        elif isinstance(ty, Function):
+            return ty.type()
+
+        elif isinstance(ty, bn.Type):
+            return self._convert_bn_type(ty)
+
+        if not ty:
```
Comment (pgoodman) [inline python/anvill/binja.py:417]: elif

File: python/anvill/binja.py [hunk 14]
```diff
@@ -173,199 +173,251 @@ def _convert_bn_llil_type(
...
+            return VoidType()
+
+        raise UnhandledTypeException("Unrecognized type passed to `Type`.", ty)
 
 
 class CallingConvention(object):
```
Comment (pgoodman) [inline python/anvill/binja.py:423]: Remove `(object):` as that Python 2, and just `class CallingConvention:`

**Review by pgoodman — CHANGES_REQUESTED** [review]

**Comment by kumarak** [comment]
Updated handling of union types. Binary ninja recovers both struct & union as struct type and set the union flag which was wrongly used.

**Review by artemdinaburg — APPROVED** [review]

File: anvill/python/anvill/binja.py [hunk 15]
```diff
@@ -173,202 +173,247 @@ def _convert_bn_llil_type(
...
+        return MacOS()
+    elif "windows" in platform:
+        return WindowsOS()
+    else:
+        raise UnhandledOSException(
+            "Missing operating system object type for OS '{}'".format(platform)
+        )
+
+
+class TypeCache:
```
[the remark this answers is not in the document]
Comment (kumarak) [reply anvill/python/anvill/binja.py:207]: I updated the comments about `TypeCache`. It provides API to convert bn types into anvill type instances recursively. It does maintain a cache of visited bn types at the program level and may have improvements in lookup time. Previously the caching was locally for each type. 

The main goal here is to refactor the existing type conversion, add global cache, and re-use stuff as much as possible.

File: anvill/python/anvill/binja.py [hunk 16]
```diff
@@ -173,202 +173,253 @@ def _convert_bn_llil_type(
...
+
+    def _convert_struct(self, tinfo: bn.types.Type):
+        """Convert bn struct type into a `Type` instance"""
+
+        assert tinfo.type_class == bn.TypeClass.StructureTypeClass
+
+        if tinfo.structure.type == bn.StructureType.UnionStructureType:
+            return self._convert_union(tinfo)
+
+        assert tinfo.structure.type == bn.StructureType.StructStructureType
```
Comment (pgoodman) [inline anvill/python/anvill/binja.py:238]: update assert to check for `ClassStructureType`

File: anvill/python/anvill/binja.py [hunk 17]
```diff
@@ -173,202 +173,253 @@ def _convert_bn_llil_type(
...
+
+    def __init__(self, bv):
+        self._bv = bv
+        self._cache = dict()
+
+    def _cache_key(self, tinfo: bn.types.Type):
+        """ Convert bn Type instance to cache key"""
+        return str(tinfo)
+
+    def _convert_struct(self, tinfo: bn.types.Type):
```
Comment (pgoodman) [inline anvill/python/anvill/binja.py:230]: Add ` -> Type` return type hints to all relevant methods.

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: anvill/python/anvill/binja.py (hunk 16 above)
Comment (kumarak) [reply anvill/python/anvill/binja.py]: Thanks, for pointing. `ClassStructureType` was not getting handled. It will be the same as `StructType()`

**Review by pgoodman — APPROVED** [review]

**Pull request merged**